README.md

<h1 align="center">Circular Seek Bar</h1>

<p align="center">
  <a href="https://flutter.dev">
    <img src="https://img.shields.io/badge/Platform-Flutter-02569B?logo=flutter"
      alt="Platform" />
  </a>

  <a href="https://pub.dartlang.org/packages/circular_seek_bar">
    <img src="https://img.shields.io/pub/v/circular_seek_bar.svg"
      alt="Pub Package" />
  </a>

  <a href="https://github.com/seosh817/circular_seek_bar/actions/workflows/main.yml">
    <img src="https://img.shields.io/github/actions/workflow/status/seosh817/circular_seek_bar/main.yml?branch=release/1.0.0&logo=github"
      alt="Build Status" />
  </a>

  <a href="https://opensource.org/licenses/MIT">
    <img src="https://img.shields.io/github/license/seosh817/Flutter_CircularSeekBar"
      alt="License: MIT" />
  </a>
</p><br>

<p align="center">Circular progress/seek bar package for flutter that supports customizable animations, dashes, and gradients.</p>

# Getting Started
- [Youtube Demo Video](#youtube-demo-video)
- [Basic Examples](#basic-examples)
  - [Basic SeekBar](#1-basic-seekbar)
  - [Gradient SeekBar](#2-gradient-seekbar)
  - [Dashed SeekBar](#3-dashed-seekbar)
  - [Add ValueNotifier](#4-add-valuenotifier)
  - [Animations](#5-animations)
- [Installing](#installing)
  - [Depend on it](#1-depend-on-it)
  - [Install it](#2-install-it)
  - [Import it](#3-import-it)
- [Properties](#properties)
- [License](#license)
- [Contribution](#contribution)

# Youtube Demo Video
An example project can be found in the [example directory](https://github.com/seosh817/Flutter_CircularSeekBar/tree/master/example) of this repository.

[![Demo Video](https://img.youtube.com/vi/rQ_-iKXHR-M/hqdefault.jpg)](https://youtu.be/rQ_-iKXHR-M)

# Basic Examples

### 1. Basic SeekBar

<img src="https://github.com/seosh817/Flutter_CircularSeekBar/blob/release/1.0.0/images/start_90_sweep_360.gif?raw=true" width="300"/>

### Dart code:

If you want to use thumb, set the thumb property values.

```dart
CircularSeekBar(
  width: double.infinity,
  height: 250,
  progress: _progress,
  barWidth: 8,
  startAngle: 90,
  sweepAngle: 360,
  strokeCap: StrokeCap.butt,
  progressColor: Colors.blue,
  innerThumbRadius: 5,
  innerThumbStrokeWidth: 3,
  innerThumbColor: Colors.white,
  outerThumbRadius: 5,
  outerThumbStrokeWidth: 10,
  outerThumbColor: Colors.blueAccent,
  animation: true,
)
```
### 2. Gradient SeekBar

<img src="https://github.com/seosh817/Flutter_CircularSeekBar/blob/release/1.0.0/images/start_45_sweep_270_gradient_rainbow.gif?raw=true" width="300"/>

```dart
CircularSeekBar(
  width: double.infinity,
  height: 250,
  progress: _progress,
  barWidth: 8,
  startAngle: 45,
  sweepAngle: 270,
  strokeCap: StrokeCap.round,
  progressGradientColors: const [Colors.red, Colors.orange, Colors.yellow, Colors.green, Colors.blue, Colors.indigo, Colors.purple],
  innerThumbRadius: 5,
  innerThumbStrokeWidth: 3,
  innerThumbColor: Colors.white,
  outerThumbRadius: 5,
  outerThumbStrokeWidth: 10,
  outerThumbColor: Colors.blueAccent,
  animation: true,
)
```
### 3. Dashed SeekBar

<img src="https://github.com/seosh817/Flutter_CircularSeekBar/blob/release/1.0.0/images/start_90_sweep_180_dashWidth_50_dashGap_15.gif?raw=true" width="300"/>


### Dart code:
```dart
CircularSeekBar(
  width: double.infinity,
  height: 250,
  progress: _progress,
  barWidth: 8,
  startAngle: 90,
  sweepAngle: 180,
  strokeCap: StrokeCap.round,
  progressGradientColors: const [Colors.blue, Colors.indigo, Colors.purple],
  dashWidth: 50,
  dashGap: 15,
  animation: true,
)
```

<img src="https://github.com/seosh817/Flutter_CircularSeekBar/blob/release/1.0.0/images/start_45_sweep_270_dashWidth_80_dashGap_15.gif?raw=true" width="300"/>

### Dart code:
```dart
CircularSeekBar(
  width: double.infinity,
  height: 250,
  progress: _progress,
  barWidth: 8,
  startAngle: 45,
  sweepAngle: 270,
  strokeCap: StrokeCap.round,
  progressGradientColors: const [Colors.blue, Colors.indigo, Colors.purple],
  dashWidth: 80,
  dashGap: 15,
  animation: true,
)
```


<img src="https://github.com/seosh817/Flutter_CircularSeekBar/blob/release/1.0.0/images/start_45_sweep_270_dashWidth_1_dashGap_2.gif?raw=true" width="300"/>

### Dart code:
```dart
CircularSeekBar(
  width: double.infinity,
  height: 250,
  progress: _progress,
  barWidth: 8,
  startAngle: 45,
  sweepAngle: 270,
  strokeCap: StrokeCap.butt,
  progressGradientColors: const [Colors.blue, Colors.indigo, Colors.purple],
  dashWidth: 1,
  dashGap: 2,
  animation: true,
)
```

### 4. Add ValueNotifier

<img src="https://github.com/seosh817/Flutter_CircularSeekBar/blob/release/1.0.0/images/start_45_sweep_270_dashWith_1_dashGap_2_gradient_rainbow.gif?raw=true" width="300"/>

### Dart code:
```dart
final ValueNotifier<double> _valueNotifier = ValueNotifier(0);
```
If you want to listen to the value of progress, create a valueNotifier and add it to the constructor.
```dart
CircularSeekBar(
  width: double.infinity,
  height: 250,
  progress: _progress,
  barWidth: 8,
  startAngle: 45,
  sweepAngle: 270,
  strokeCap: StrokeCap.butt,
  progressGradientColors: const [Colors.red, Colors.orange, Colors.yellow, Colors.green, Colors.blue, Colors.indigo, Colors.purple],
  innerThumbRadius: 5,
  innerThumbStrokeWidth: 3,
  innerThumbColor: Colors.white,
  outerThumbRadius: 5,
  outerThumbStrokeWidth: 10,
  outerThumbColor: Colors.blueAccent,
  dashWidth: 1,
  dashGap: 2,
  animation: true,
  valueNotifier: _valueNotifier,
  child: Center(
    child: ValueListenableBuilder(
        valueListenable: _valueNotifier,
        builder: (_, double value, __) => Column(
              mainAxisSize: MainAxisSize.min,
              children: [
                Text('${value.round()}', style: kNotoSansBold16.copyWith(color: Colors.white)),
                Text('progress', style: kNotoSansRegular14.copyWith(color: Colors.grey)),
              ],
            )),
  ),
)
```

### 5. Animations

<img src="https://github.com/seosh817/Flutter_CircularSeekBar/blob/hotfix/1.0.3/images/animations_bounceOut.gif?raw=true" width="300"/>


### Dart code:

Various animations can be applied to the SeekBar by changing the `curves` property.

```dart
CircularSeekBar(
  width: double.infinity,
  height: 250,
  progress: _progress,
  barWidth: 8,
  startAngle: 45,
  sweepAngle: 270,
  strokeCap: StrokeCap.butt,
  progressGradientColors: const [Colors.red, Colors.orange, Colors.yellow, Colors.green, Colors.blue, Colors.indigo, Colors.purple],
  innerThumbRadius: 5,
  innerThumbStrokeWidth: 3,
  innerThumbColor: Colors.white,
  outerThumbRadius: 5,
  outerThumbStrokeWidth: 10,
  outerThumbColor: Colors.blueAccent,
  dashWidth: 1,
  dashGap: 2,
  animation: true,
  curves: Curves.bounceOut,
  valueNotifier: _valueNotifier,
  child: Center(
    child: ValueListenableBuilder(
        valueListenable: _valueNotifier,
        builder: (_, double value, __) => Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Text('${value.round()}', style: kNotoSansBold16.copyWith(color: Colors.white)),
            Text('progress', style: kNotoSansRegular14.copyWith(color: Colors.grey)),
          ],
        )),
  ),
)
```

# Installing

### 1. Depend on it

Add this to your package's `pubspec.yaml` file:

```yaml
dependencies:
  circular_seek_bar: ^1.1.0
```

or

```
$ flutter pub add circular_seek_bar
```

### 2. Install it

You can install packages from the command line:

```
$ flutter pub get
```

### 3. Import it

Now in your `Dart` code, you can use:

```dart
import 'package:circular_seek_bar/circular_seek_bar.dart';
```

## Properties

You can customize the CircularSeekBar using the following properties:

|Property|Type|Default|Description|
|:---|:---|:---|:---|
| width | `double` | required | CircularSeekBar width.|
| height | `double` | required | CircularSeekBar height.|
| progress | `double` | 0 | Current value of seek bar. |
| minProgress | `double` | 0 | Minimum value of seek bar.|
| maxProgress | `double` | 100 | Maximum value of seek bar.|
| startAngle | `double` | 0 | The Angle to start drawing this seek bar from.|
| sweepAngle | `double` | 360 | The Angle through which to draw the seek bar.|
| barWidth | `double` | 10 | The thickness of the seek bar.|
| trackColor | `Color` | Colors.white54 | Background track color of seek bar.|
| trackGradientColors | `List<Color>` | [] | Background track gradient colors of seek bar.<br>If [trackGradientColors] is not empty, [trackColor] is not applied.|
| progressColor | `Color` | Colors.blue | Foreground progress color of seek bar.|
| progressGradientColors | `List<Color>` | [] | Foreground progressGradientColors of seek bar.<br>If [progressGradientColors] is not empty, [progressColor] is not applied.|
| strokeCap | `StrokeCap` | StrokeCap.round | Styles to use for arcs endings.|
| animation | `bool` | true | Active seek bar animation.|
| animDurationMillis | `int` | 1000 | Animation duration milliseconds.|
| curves | `Curve` | Curves.linear | Animation curve.|
| innerThumbRadius | `double` | 0 | The radius of the seekbar inner thumb.|
| innerThumbStrokeWidth | `double` | 0 | The stroke width of the seekbar inner thumb.|
| innerThumbColor | `Color` | Colors.white | Color of the seekbar inner thumb.|
| outerThumbRadius | `double` | 0 | The radius of the seekbar outer thumb.|
| outerThumbStrokeWidth | `double` | 0 | The stroke width of the seekbar outer thumb.|
| outerThumbColor | `Color` | Colors.white | Color of the seekbar outer thumb.|
| dashWidth | `double` | 0 | Dash width of seek bar.|
| dashGap | `double` | 0 |  Dash gap of seek bar.|
| valueNotifier | `ValueNotifier<double>?` | null | This ValueNotifier notifies the listener that the seekbar's progress value has changed.|
| onEnd | `VoidCallback?` | null | This callback function will execute when Animation is finished.|
| interactive | `bool` | true | Set to true if you want to interact with TapDown to change the seekbar's progress.|
| child | `Widget?` | null | This widget is placed on the seek bar.|

# License
```
MIT License

Copyright (c) 2022 seosh817

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
```

# Contribution
Feel free to file an [issue](https://github.com/seosh817/Flutter_CircularSeekBar/issues) if you find a problem or make [pull requests](https://github.com/seosh817/Flutter_CircularSeekBar/pulls).<br>
All contributions are welcome 😁



CHANGELOG.md
## 1.0.0

- Initial release

## 1.0.0+1

- Edit README.md and description

## 1.0.1

- reformat dart code and edit README.md

## 1.0.2

- Fix bug: Fix progress not updated bug

## 1.0.3

- Fix bug: Fix progress not updated bug
- Add animation description 

## 1.0.3+1

- Edit description

## 1.0.3+2

- Change GitHub workflow badge routes

## 1.0.3+3

- Change GitHub workflow badge routes

## 1.1.0

- Modify the formula to get the radius of CircularSeekBar

- Fix bug: Large barWidth with strokeCap round looks odd(#1)

## 1.1.0+1

- Format dart code

## 1.1.0+2

- Update for pub score

## 1.1.1

- Upgrade supported flutter and dart version 

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:circular_seek_bar_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/circular_seek_bar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:circular_seek_bar/circular_seek_bar.dart';

void main() {
  test('test circularSeekBar', () {
    const circularSeekBar = CircularSeekBar(
      width: 300,
      height: 300,
      progress: 50,
    );
    expect(circularSeekBar.progress, 50);
  });
}


example/lib/util/text_style.dart
import 'package:flutter/material.dart';

const TextStyle kNotoSansBold12 = TextStyle(
    fontFamily: 'NotoSans', fontWeight: FontWeight.w700, fontSize: 12);
const TextStyle kNotoSansRegular12 =
    TextStyle(fontFamily: 'NotoSans', fontSize: 12);
const TextStyle kNotoSansMedium12 = TextStyle(
    fontFamily: 'NotoSans', fontWeight: FontWeight.w500, fontSize: 12);

const TextStyle kNotoSansBold14 = TextStyle(
    fontFamily: 'NotoSans', fontWeight: FontWeight.w700, fontSize: 14);
const TextStyle kNotoSansRegular14 =
    TextStyle(fontFamily: 'NotoSans', fontSize: 14);
const TextStyle kNotoSansMedium14 = TextStyle(
    fontFamily: 'NotoSans', fontWeight: FontWeight.w500, fontSize: 14);

const TextStyle kNotoSansBold16 = TextStyle(
    fontFamily: 'NotoSans', fontWeight: FontWeight.w700, fontSize: 16);
const TextStyle kNotoSansRegular16 =
    TextStyle(fontFamily: 'NotoSans', fontSize: 16);
const TextStyle kNotoSansMedium16 = TextStyle(
    fontFamily: 'NotoSans', fontWeight: FontWeight.w500, fontSize: 16);

const TextStyle kNotoSansBold24 = TextStyle(
    fontFamily: 'NotoSans', fontWeight: FontWeight.w700, fontSize: 24);
const TextStyle kNotoSansRegular24 =
    TextStyle(fontFamily: 'NotoSans', fontSize: 24);
const TextStyle kNotoSansMedium24 = TextStyle(
    fontFamily: 'NotoSans', fontWeight: FontWeight.w500, fontSize: 24);


example/lib/main.dart
import 'package:circular_seek_bar/circular_seek_bar.dart';
import 'package:circular_seek_bar_example/util/text_style.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter circular_seek_bar example',
      theme: ThemeData(
        primaryColor: Colors.blue,
        brightness: Brightness.dark,
      ),
      home: const CircularSeekBarExamplePage(),
    );
  }
}

class CircularSeekBarExamplePage extends StatefulWidget {
  const CircularSeekBarExamplePage({Key? key}) : super(key: key);

  @override
  State<CircularSeekBarExamplePage> createState() =>
      _CircularSeekBarExamplePageState();
}

class _CircularSeekBarExamplePageState
    extends State<CircularSeekBarExamplePage> {
  final ValueNotifier<double> _valueNotifier = ValueNotifier(0);

  double _progress = 90;
  double _startAngle = 45;
  double _sweepAngle = 270;
  double _dashWidth = 0;
  double _dashGap = 0;
  double _barWidth = 8.0;
  bool _useGradient = true;
  bool _rounded = true;
  bool _animation = true;
  bool _thumbVisible = true;
  bool _interactive = true;

  @override
  void dispose() {
    _valueNotifier.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Flutter circular_seek_bar example'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.start,
          children: <Widget>[
            CircularSeekBar(
              width: double.infinity,
              progress: _progress,
              height: 250,
              barWidth: _barWidth,
              startAngle: _startAngle,
              sweepAngle: _sweepAngle,
              strokeCap: _rounded ? StrokeCap.round : StrokeCap.butt,
              progressGradientColors: _useGradient
                  ? [
                      Colors.red,
                      Colors.orange,
                      Colors.yellow,
                      Colors.green,
                      Colors.blue,
                      Colors.indigo,
                      Colors.purple
                    ]
                  : [],
              dashWidth: _dashWidth,
              dashGap: _dashGap,
              animation: _animation,
              curves: Curves.linear,
              innerThumbRadius: _thumbVisible ? 5 : 0,
              innerThumbStrokeWidth: _thumbVisible ? 3 : 0,
              outerThumbRadius: _thumbVisible ? 5 : 0,
              outerThumbStrokeWidth: _thumbVisible ? 10 : 0,
              valueNotifier: _valueNotifier,
              interactive: _interactive,
              child: Center(
                child: ValueListenableBuilder(
                    valueListenable: _valueNotifier,
                    builder: (_, double value, __) => Column(
                          mainAxisSize: MainAxisSize.min,
                          children: [
                            Text('${value.round()}',
                                style: kNotoSansBold16.copyWith(
                                    color: Colors.white)),
                            Text('progress',
                                style: kNotoSansRegular14.copyWith(
                                    color: Colors.grey)),
                          ],
                        )),
              ),
            ),
            const SizedBox(
              height: 20,
            ),
            Expanded(
              child: SingleChildScrollView(
                child: Column(
                  children: [
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 16.0),
                      child: Row(
                        children: [
                          Text(
                            'progress: ',
                            style:
                                kNotoSansBold14.copyWith(color: Colors.white),
                          ),
                          Slider(
                            min: 0.0,
                            max: 100.0,
                            value: _progress,
                            onChanged: (value) {
                              setState(() {
                                _progress = value;
                              });
                            },
                          ),
                        ],
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 16.0),
                      child: Row(
                        children: [
                          Text(
                            'startAngle: ',
                            style:
                                kNotoSansBold14.copyWith(color: Colors.white),
                          ),
                          Slider(
                            min: 0.0,
                            max: 360.0,
                            value: _startAngle,
                            onChanged: (value) {
                              setState(() {
                                _startAngle = value;
                              });
                            },
                          ),
                        ],
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 16.0),
                      child: Row(
                        children: [
                          Text(
                            'sweepAngle: ',
                            style:
                                kNotoSansBold14.copyWith(color: Colors.white),
                          ),
                          Slider(
                            min: 0.0,
                            max: 360.0,
                            value: _sweepAngle,
                            onChanged: (value) {
                              setState(() {
                                _sweepAngle = value;
                              });
                            },
                          ),
                        ],
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 16.0),
                      child: Row(
                        children: [
                          Text(
                            'dashWidth: ',
                            style:
                                kNotoSansBold14.copyWith(color: Colors.white),
                          ),
                          Slider(
                            min: 0.0,
                            max: 40.0,
                            value: _dashWidth,
                            onChanged: (value) {
                              setState(() {
                                _dashWidth = value;
                              });
                            },
                          ),
                        ],
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 16.0),
                      child: Row(
                        children: [
                          Text(
                            'dashGap: ',
                            style:
                                kNotoSansBold14.copyWith(color: Colors.white),
                          ),
                          Slider(
                            min: 0.0,
                            max: 10.0,
                            value: _dashGap,
                            onChanged: (value) {
                              setState(() {
                                _dashGap = value;
                              });
                            },
                          ),
                        ],
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 16.0),
                      child: Row(
                        children: [
                          Text(
                            'barWidth: ',
                            style:
                                kNotoSansBold14.copyWith(color: Colors.white),
                          ),
                          Slider(
                            min: 1.0,
                            max: 10,
                            value: _barWidth,
                            onChanged: (value) {
                              setState(() {
                                _barWidth = value;
                              });
                            },
                          ),
                        ],
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 16.0),
                      child: Row(
                        children: [
                          Text(
                            'useGradient: ',
                            style:
                                kNotoSansBold14.copyWith(color: Colors.white),
                          ),
                          Switch(
                              value: _useGradient,
                              onChanged: (value) {
                                setState(() {
                                  _useGradient = value;
                                });
                              }),
                          Text(
                            'rounded: ',
                            style:
                                kNotoSansBold14.copyWith(color: Colors.white),
                          ),
                          Switch(
                              value: _rounded,
                              onChanged: (value) {
                                setState(() {
                                  _rounded = value;
                                });
                              }),
                        ],
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 16.0),
                      child: Row(
                        children: [
                          Text(
                            'animation: ',
                            style:
                                kNotoSansBold14.copyWith(color: Colors.white),
                          ),
                          Switch(
                              value: _animation,
                              onChanged: (value) {
                                setState(() {
                                  _animation = value;
                                });
                              }),
                          Text(
                            'thumbVisible: ',
                            style:
                                kNotoSansBold14.copyWith(color: Colors.white),
                          ),
                          Switch(
                              value: _thumbVisible,
                              onChanged: (value) {
                                setState(() {
                                  _thumbVisible = value;
                                });
                              }),
                        ],
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 16.0),
                      child: Row(
                        children: [
                          Text(
                            'interactive: ',
                            style:
                                kNotoSansBold14.copyWith(color: Colors.white),
                          ),
                          Switch(
                              value: _interactive,
                              onChanged: (value) {
                                setState(() {
                                  _interactive = value;
                                });
                              }),
                        ],
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
